### ISSUE TITLE
Category -> Product Media Images not shown (Heisenbug)

### ISSUE THREAD

**Issue by ulftietze** [magento/magento2#17900]
### Preconditions

1. Magento 2.2.3 (But we observed this since >= 2.1.2)
2. PHP 7
3. Flat Product catalog enabled
4. One Store with 2 Websites with each of them simple StoreView
5. One Nginx as Loadbalancer, 1 Varnish, 2-4 Frontend Nodes, 2 Shared Data nfs, 4 server DatabaseCluster
6. Indexer update on schedule
7. Cron runs every minute (like Magento wants it to)

### Steps to reproduce

This is kinda difficult to explain. It not happens that often and reproducable as i want to :/
I try to say a few words about how we can recognize this or what happend before we record that misbehaviour.
1. Happens after we do an import of products (save over rest api, import over csv files)
2. This import files just include qty and/or status (sometimes all data, but we can reproduce this with just status and qty)
2. The indexer runs and may get a deadlock here
3. The indexer cron: "indexer_reindex_all_invalid" often fails with a ``` SQLSTATE[HY000]: General error: 1205 Lock wait timeout exceeded; try restarting transaction, query was:  ... ``` 

### Expected result

1. Indexer runs probably and nice and updates all the linked small images and stuff
2. Category Page has it's product images back.

### Actual result

1. The indexer does not update it's database structure correctly
2. We have missing information in the flat product tables
```
mysql> select * from catalog_product_flat_1 where catalog_product_flat_1.sku = "produktsku"\G
*************************** 1. row ***************************
                     entity_id: 13355
              attribute_set_id: 19
                       type_id: simple
                         color: NULL
                   color_value: NULL
                  color_filter: 1880
                    created_at: 2018-06-08 11:02:03
                 delivery_time: NULL
                   description: NULL
                           ean: NULL
        gift_message_available: NULL
                   has_options: 0
                         image: no_selection
                   image_label: NULL
                        is_hot: NULL
                   links_exist: NULL
    links_purchased_separately: NULL
                   links_title: NULL
                          msrp: NULL
msrp_display_actual_price_type: NULL
                          name: Produktname
    neusta_product_active_from: 2018-06-25 00:00:00
      neusta_product_active_to: NULL
             newest_sort_order: 0
                news_from_date: NULL
                  news_to_date: NULL
            popular_sort_order: NULL
                         price: 1.9500
                    price_type: NULL
                    price_view: NULL
              required_options: 0
             short_description: NULL
                           sku: produktsku
                      sku_type: NULL
                   small_image: no_selection
             small_image_label: NULL
             special_from_date: 2018-06-08 11:02:03
                 special_price: 1.9500
               special_to_date: NULL
                  swatch_image: no_selection
                  tax_class_id: NULL
      tel_required_in_checkout: NULL
                     thumbnail: no_selection
               thumbnail_label: NULL
                    updated_at: 2018-09-03 13:52:18
                       url_key: producturlkey
                      url_path: NULL
                    visibility: 4
                        weight: 260.0000
                   weight_type: NULL
1 row in set (0.02 sec)
```

May you have some information about this problem. Sorry that i can't write more stuff here :/

Greetz, Ulf

**Comment by magento-engcom-team** [comment]
Hi @ulftietze. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me {$VERSION} instance` 

where `{$VERSION}` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (2.2-develop +). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@ulftietze do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by orlangur** [comment]
@ulftietze flat tables are not recommended anymore as they do not give significant performance boost.

**Comment by ghost** [comment]
Hi @ulftietze  seems like this are duplicat for https://github.com/magento/magento2/issues/8933

**Comment by hostep** [comment]
@orlangur:

> flat tables are not recommended anymore as they do not give significant performance boost.

This is dependent on a project per project base, we have projects running the latest Magento version which still show some performance benefits when flat tables are enabled (tested with FPC disabled of course).
So I don't think this statement can be generalised.

But we also had to disable the flats on one specific project to circumvent this deadlock problem.

**Comment by ulftietze** [comment]
Hey Guys,

thanks for referencing this issue @engcom-backlog-nazar :)
This may be a bug which results from the deadlock error. Hope this gets fixed in near future ^^

I dunno if this brings really some performance boosts in our case, it looks like (if flat products are disabled) it's a bit faster now. 
If we take a deeper look into this, i try to report my knowledgements.

Thanks and greets,
Ulf

**Comment by ghost** [comment]
@ulftietze Can i close ticket as duplicate? they are similar and have identical sense. And you can reopen if you have any updates.

**Comment by ulftietze** [comment]
@engcom-backlog-nazar 
If you promise me, that this issue is going to be fixed with the #8933, then yes :laughing:

**Comment by ghost** [comment]
@ulftietze I close it , as we can just reopen it again in future if someone else finds it. If someone fixed references issue,  it may fix yours issue.
